### PR TITLE
Add throughput data store to ServiceControl RavenDB persistence

### PR DESCRIPTION
**Pull request by abparticular**

### CODE REVIEW

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    }
+
+    public async Task<IEnumerable<(EndpointIdentifier, Endpoint?)>> GetEndpoints(IEnumerable<EndpointIdentifier> endpointIds, CancellationToken cancellationToken)
+    {
+        var endpoints = new List<(EndpointIdentifier, Endpoint?)>();
+
+        var store = await storeProvider.GetDocumentStore(cancellationToken);
+        using IAsyncDocumentSession session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        var documentIdLookup = endpointIds.ToDictionary(
```
Comment (johnsimons) [inline src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs:78]: Possible multiple enumeration of `endpointIds` ?

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        var store = await storeProvider.GetDocumentStore(cancellationToken);
+        using IAsyncDocumentSession session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        await session.StoreAsync(document, document.GenerateDocumentId(), cancellationToken);
+        await session.SaveChangesAsync(cancellationToken);
+    }
+
+    public async Task<IDictionary<string, IEnumerable<ThroughputData>>> GetEndpointThroughputByQueueName(IEnumerable<string> queueNames, CancellationToken cancellationToken)
+    {
+        var results = queueNames.ToDictionary(queueName => queueName, queueNames => new List<ThroughputData>() as IEnumerable<ThroughputData>);
```
Comment (johnsimons) [inline src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs:122]: Possible multiple enumeration of `queueNames` ?

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+        var store = await storeProvider.GetDocumentStore(cancellationToken);
+        using IAsyncDocumentSession session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        var query = session.Query<EndpointDocument>()
+            .Where(document => document.SanitizedName.In(queueNames))
+            .Include(builder => builder.IncludeTimeSeries(ThroughputTimeSeriesName));
+
+        var documents = await query.ToListAsync(cancellationToken);
```
Comment (johnsimons) [inline src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs:131]: Should this stream the results?

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                throughputDataList.Add(throughputData);
+            }
+        }
+
+        return results;
+    }
+
+    public async Task RecordEndpointThroughput(string endpointName, ThroughputSource throughputSource, IEnumerable<EndpointDailyThroughput> throughput, CancellationToken cancellationToken)
+    {
+        if (!throughput.Any())
```
Comment (johnsimons) [inline src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs:156]: Possible multiple enumeration of  `throughput` ?

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    public async Task UpdateUserIndicatorOnEndpoints(List<UpdateUserIndicator> userIndicatorUpdates, CancellationToken cancellationToken)
+    {
+        var updates = userIndicatorUpdates.ToDictionary(u => u.Name, u => u.UserIndicator);
+        var store = await storeProvider.GetDocumentStore(cancellationToken);
+        using IAsyncDocumentSession session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        var query = session.Query<EndpointDocument>()
+            .Where(document => document.SanitizedName.In(updates.Keys));
+
+        var documents = await query.ToListAsync(cancellationToken);
```
Comment (johnsimons) [inline src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs:188]: Should this stream?

**Review by johnsimons — APPROVED** [review]
LGTM, added a few comments for consideration

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs (hunk 1 above)
Comment (abparticular) [reply src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs]: Fixed

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs (hunk 2 above)
Comment (abparticular) [reply src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs]: Fixed

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs (hunk 4 above)
Comment (abparticular) [reply src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs]: Fixed

File: src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs (hunk 3 above)
Comment (abparticular) [reply src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs]: This implementation does not do streaming, it is a possible future enhancement.

**Comment by johnsimons** [comment]
We are missing the tests in this PR

**Review by jpalac — APPROVED** [review]

File: src/ServiceControl.Persistence.RavenDB/ServiceControl.Persistence.RavenDB.csproj [hunk 6]
```diff
@@ -8,6 +8,8 @@
 
   <ItemGroup>
     <!-- Private=false & ExcludeAssets=runtime prevent repeatedly including binary dependencies of ServiceControl.Persistence and its depenencies in each persister directory -->
+    <ProjectReference Include="..\Particular.LicensingComponent.Contracts\Particular.LicensingComponent.Contracts.csproj" />
```
Comment (johnsimons) [inline src/ServiceControl.Persistence.RavenDB/ServiceControl.Persistence.RavenDB.csproj:11]:
```suggestion
    <ProjectReference Include="..\Particular.LicensingComponent.Contracts\Particular.LicensingComponent.Contracts.csproj"  Private="false" ExcludeAssets="runtime" />
```

**Review by johnsimons — CHANGES_REQUESTED** [review]

**Review by jpalac — APPROVED** [review]

**Pull request merged**